### ISSUE TITLE
Error: Cannot find module

### ISSUE THREAD

**Issue by Mockapapella** [Seo-Rii/electron-acrylic-window#46]
# Problem
I'm having an issue building `electron-acrylic-window` with file `win_delay_load_hook.cc`

# My setup: 
- Windows 10 Enterprise LTSC, Version 10.0.17763 Build 17763
- node v12.19.0
- node-gyp v7.1.2
- Python 2.7
- tsc Version 4.1.3

I've made sure to install VS Build Tools 2017 and 2019 and switching the configuration between the two using `npm config set msvs_version 201<x>`. I tried that installing VS Build Tools/Python manually and with `npm install --global windows-build-tools`

# Logs
This is what shows up when I try to install `electron-acrylic-window` via `npm install electron-acrylic-window`:

```
C:\Software_Development\Personal\JavaScript\ray-view>npm install electron-acrylic-window

> electron-acrylic-window@0.5.4 install C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window
> (node-gyp rebuild && npm run-script build-ts)


C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
  win10.cpp
  win_delay_load_hook.cc
c:\software_development\personal\javascript\ray-view\node_modules\electron-acrylic-window\src\native\win10.cpp(38): warning C4530: C++ exception handler used, but un
wind semantics are not enabled. Specify /EHsc [C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\vibrancy-wrapper.vcxpr
oj]
c:\software_development\personal\javascript\ray-view\node_modules\electron-acrylic-window\src\native\win10.cpp(77): warning C4101: 'ex': unreferenced local variable 
[C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\vibrancy-wrapper.vcxproj]
c:\software_development\personal\javascript\ray-view\node_modules\electron-acrylic-window\src\native\win10.cpp(110): warning C4101: 'ex': unreferenced local variable
 [C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\vibrancy-wrapper.vcxproj]
     Creating library C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\Release\vibrancy-wrapper.lib and object C:\Soft
  ware_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\Release\vibrancy-wrapper.exp
  vibrancy-wrapper.vcxproj -> C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\Release\\vibrancy-wrapper.node

> electron-acrylic-window@0.5.4 build-ts C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window
> tsc

npm WARN ray-view@1.0.0 No repository field.

+ electron-acrylic-window@0.5.4
added 11 packages from 77 contributors and audited 578 packages in 16.312s

43 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

As a result of this error (presumably, from what I can tell), no `index.js` file is every generated in `electron-acrylic-window`'s `build/` folder. When I run my code...

```
const { app, BrowserWindow } = require("electron");
const path = require("path");
const { setVibrancy } = require("electron-acrylic-window");

// Handle creating/removing shortcuts on Windows when installing/uninstalling.
if (require("electron-squirrel-startup")) {
  // eslint-disable-line global-require
  app.quit();
}

const createWindow = () => {
  // Create the browser window.
  const mainWindow = new BrowserWindow({
    width: 800,
    height: 600,
    frame: false,
    webPreferences: {
      nodeIntegration: true,
      enableRemoteModule: true,
    },
  });
  mainWindow.loadFile(path.join(__dirname, "index.html"));
  mainWindow.webContents.openDevTools();

  // Vibrancy Settings
  vibrancy_options = {
    theme: "#661237cc",
    effect: "acrylic",
    useCustomWindowRefreshMethod: true,
    disableOnBlur: true,
    debug: false,
  };
  setVibrancy(mainWindow, vibrancy_options);
  mainWindow.show();
};
app.on("ready", createWindow);
app.on("window-all-closed", () => {
  if (process.platform !== "darwin") {
    app.quit();
  }
});

app.on("activate", () => {
  if (BrowserWindow.getAllWindows().length === 0) {
    createWindow();
  }
});
```

...this error message pops up:
```
> ray-view@1.0.0 start C:\Software_Development\Personal\JavaScript\ray-view
> electron-forge start

√ Checking your system
√ Locating Application
- Preparing native dependencies: 0 / 1Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
/ Preparing native dependencies: 0 / 1  win10.cpp
- Preparing native dependencies: 0 / 1c:\software_development\personal\javascript\ray-view\node_modules\electron-acrylic-window\src\native\win10.cpp(38): warning C4530: C++ exception handler used, but un
wind semantics are not enabled. Specify /EHsc [C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\vibrancy-wrapper.vcxpr
oj]
c:\software_development\personal\javascript\ray-view\node_modules\electron-acrylic-window\src\native\win10.cpp(77): warning C4101: 'ex': unreferenced local variable 
[C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\vibrancy-wrapper.vcxproj]
c:\software_development\personal\javascript\ray-view\node_modules\electron-acrylic-window\src\native\win10.cpp(110): warning C4101: 'ex': unreferenced local variable
 [C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\vibrancy-wrapper.vcxproj]
| Preparing native dependencies: 0 / 1  win_delay_load_hook.cc
\ Preparing native dependencies: 0 / 1     Creating library C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\Release\vibrancy-wrapper.lib and object C:\Soft
  ware_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\Release\vibrancy-wrapper.exp
  Generating code
| Preparing native dependencies: 0 / 1  All 102 functions were compiled because no usable IPDB/IOBJ from previous compilation was found.
  Finished generating code
- Preparing native dependencies: 0 / 1  vibrancy-wrapper.vcxproj -> C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\Release\\vibrancy-wrapper.node
- Preparing native dependencies: 1 / 2Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
- Preparing native dependencies: 1 / 2  win32-displayconfig.cc
- Preparing native dependencies: 1 / 2  win_delay_load_hook.cc
/ Preparing native dependencies: 1 / 2     Creating library C:\Software_Development\Personal\JavaScript\ray-view\node_modules\win32-displayconfig\build\Release\win32_displayconfig.lib and object C:\Softw
  are_Development\Personal\JavaScript\ray-view\node_modules\win32-displayconfig\build\Release\win32_displayconfig.exp
  Generating code
| Preparing native dependencies: 1 / 2  All 991 functions were compiled because no usable IPDB/IOBJ from previous compilation was found.
  Finished generating code
\ Preparing native dependencies: 1 / 2  win32_displayconfig.vcxproj -> C:\Software_Development\Personal\JavaScript\ray-view\node_modules\win32-displayconfig\build\Release\\win32_displayconfig.node
√ Preparing native dependencies: 2 / 2
√ Launching Application

App threw an error during load
Error: Cannot find module 'C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\index.js'. Please verify that the package.json has a valid "main" entry
    at tryPackage (internal/modules/cjs/loader.js:328:19)
    at Function.Module._findPath (internal/modules/cjs/loader.js:689:18)
    at Module._resolveFilename (internal/modules/cjs/loader.js:959:27)
    at Function.n._resolveFilename (electron/js2c/browser_init.js:249:921)
    at Module._load (internal/modules/cjs/loader.js:848:27)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12738)
    at Module.require (internal/modules/cjs/loader.js:1032:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (C:\Software_Development\Personal\JavaScript\ray-view\src\index.js:3:25)
    at Module._compile (internal/modules/cjs/loader.js:1152:30)
```

# Misc
I've tried installing the Windows 10 SDK as mentioned in Seo-Rii's comment [here](https://github.com/Seo-Rii/electron-acrylic-window/issues/38#issuecomment-717652536), but that didn't work unfortunately.

**Comment by tscpp** [comment]
@Seo-Rii publish the package **with** the JavaScript build files. Building the TypeScript while installing the package makes additional build time and may result in errors too.

**Comment by tscpp** [comment]
@Mockapapella try building the index file manually by:
```
npm i -D electron
npm i electron-acrylic-window
cd node_modules/electron-acrylic-window
tsc -p ./tsconfig.json 
```

**Comment by Mockapapella** [comment]
I'm still getting the error when I try to run my application:

```
C:\Software_Development\Personal\JavaScript\ray-view>npm start

> ray-view@1.0.0 start C:\Software_Development\Personal\JavaScript\ray-view
> electron-forge start

√ Checking your system
√ Locating Application
\ Preparing native dependencies: 0 / 1Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
\ Preparing native dependencies: 0 / 1  win10.cpp
| Preparing native dependencies: 0 / 1C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\src\native\win10.cpp(38,9): warning C4530: C++ exception handler used, but 
unwind semantics are not enabled. Specify /EHsc [C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\vibrancy-wrapper.vcx
proj]
C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\src\native\win10.cpp(77,26): warning C4101: 'ex': unreferenced local variab
le [C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\vibrancy-wrapper.vcxproj]
C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\src\native\win10.cpp(110,26): warning C4101: 'ex': unreferenced local varia
ble [C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\vibrancy-wrapper.vcxproj]
| Preparing native dependencies: 0 / 1  win_delay_load_hook.cc
- Preparing native dependencies: 0 / 1     Creating library C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\Release\vibrancy-wrapper.lib and object C:\Soft
  ware_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\Release\vibrancy-wrapper.exp
  Generating code
  Previous IPDB not found, fall back to full compilation.
\ Preparing native dependencies: 0 / 1  All 97 functions were compiled because no usable IPDB/IOBJ from previous compilation was found.
  Finished generating code
/ Preparing native dependencies: 0 / 1  vibrancy-wrapper.vcxproj -> C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\Release\\vibrancy-wrapper.node
/ Preparing native dependencies: 1 / 2Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
/ Preparing native dependencies: 1 / 2  win32-displayconfig.cc
- Preparing native dependencies: 1 / 2  win_delay_load_hook.cc
/ Preparing native dependencies: 1 / 2     Creating library C:\Software_Development\Personal\JavaScript\ray-view\node_modules\win32-displayconfig\build\Release\win32_displayconfig.lib and object C:\Softw
  are_Development\Personal\JavaScript\ray-view\node_modules\win32-displayconfig\build\Release\win32_displayconfig.exp
  Generating code
  Previous IPDB not found, fall back to full compilation.
\ Preparing native dependencies: 1 / 2  All 846 functions were compiled because no usable IPDB/IOBJ from previous compilation was found.
  Finished generating code
/ Preparing native dependencies: 1 / 2  win32_displayconfig.vcxproj -> C:\Software_Development\Personal\JavaScript\ray-view\node_modules\win32-displayconfig\build\Release\\win32_displayconfig.node
√ Preparing native dependencies: 2 / 2
√ Launching Application

App threw an error during load
Error: Cannot find module 'C:\Software_Development\Personal\JavaScript\ray-view\node_modules\electron-acrylic-window\build\index.js'. Please verify that the package.json has a valid "main" entry
    at tryPackage (internal/modules/cjs/loader.js:328:19)
    at Function.Module._findPath (internal/modules/cjs/loader.js:689:18)
    at Module._resolveFilename (internal/modules/cjs/loader.js:959:27)
    at Function.n._resolveFilename (electron/js2c/browser_init.js:249:921)
    at Module._load (internal/modules/cjs/loader.js:848:27)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12738)
    at Module.require (internal/modules/cjs/loader.js:1032:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (C:\Software_Development\Personal\JavaScript\ray-view\src\index.js:3:25)
    at Module._compile (internal/modules/cjs/loader.js:1152:30)
```

**Comment by tscpp** [comment]
@Mockapapella would you try to make minimal reprudution of the issue.

**Comment by Mockapapella** [comment]
> @Mockapapella would you try to make minimal reproduction of the issue.

- Uninstall both electron and electron-acrylic-window (this project already had electron installed)
- Enter these commands in the terminal in my project folder line by line

```
npm i -D electron
npm i electron-acrylic-window
cd node_modules/electron-acrylic-window
tsc -p ./tsconfig.json 
```

(note that there doesn't seem to be a difference now compared to before I posted this issue)

- run `npm start` in the root of my project folder

---------------

I tried to install lower versions of electron-acrylic-window (0.5-0.5.4) that had the same problem.

**Comment by Mockapapella** [comment]
I'm not sure how relevant it is, but when I clone the repo right from GitHub and run the test application everything works flawlessly.

**Comment by Seo-Rii** [comment]
@Mockapapella Before a new version, including the typescript build file, please use the previous version without the typescript applied. Since the typescript was applied from v0.5.0, it will work if you use v0.4.7.

**Comment by Mockapapella** [comment]
![image](https://user-images.githubusercontent.com/17628762/103165547-e548d700-47de-11eb-9702-6e6d1d73ae52.png)

I'm getting this error now instead. The app starts now but it doesn't have the frosted glass look.

**Comment by tscpp** [comment]
@Mockapapella what version of electron are you using?

Try using this code before using `setVibrancy`:
```javascript
console.log(Object.keys(mainWindow).includes('getNativeWindowHandle'), Object.keys(Buffer.from('')).includes('readInt32LE'))
```

**Comment by Mockapapella** [comment]
Electron v6.14.8

```
// Vibrancy Settings
  vibrancy_options = {
    theme: "#661237cc",
    effect: "acrylic",
    useCustomWindowRefreshMethod: true,
    disableOnBlur: true,
    debug: true,
  };
  console.log("Hello World");
  console.log(
    Object.keys(mainWindow).includes("getNativeWindowHandle"),
    Object.keys(Buffer.from("")).includes("readInt32LE")
  );
  setVibrancy(mainWindow, vibrancy_options);
  mainWindow.show();
```

So your script didn't produce any output, but for some reason neither did `console.log("Hello World");`. I got the same error as in my previous comment.

**Comment by Seo-Rii** [comment]
The problem is caused by the incompatibility of the getNativeWindowHandle API used to get hWnd because the electron version is too low. Please update the Electron version.

**Comment by tscpp** [comment]
@Seo-Rii `getNativeWindowHandle` exists in electron v6.1.12 https://github.com/electron/electron/blob/e01b1831d96d5d68f54af879b00c617358df5372/docs/api/browser-window.md#wingetnativewindowhandle
As far as I can see, there is no v6.14.8 version of electron in npm or GitHub.

**Comment by tscpp** [comment]
@Mockapapella use `fs.writeFileSync` and test if that works:

```javascript
require('fs').writeFileSync('output.txt',  (Object.keys(mainWindow).includes("getNativeWindowHandle") ? '1' : '0') + (Object.keys(Buffer.from("")).includes("readInt32LE") ? '1' : '0'))
```

If it does succeed, comment the contents of output.txt.

**Comment by Mockapapella** [comment]
> As far as I can see, there is no v6.14.8 version of electron in npm or GitHub.

Whoops, my bad. That's my npm version. My electron version is 11.1.1.

@tscpp The output of that script is `00`

**Comment by tscpp** [comment]
@Mockapapella try using:

```
require('fs').writeFileSync('output.txt', ('getNativeWindowHandle' in win).toString())
```

I forgot Object.keys cannot extract from classes.

**Comment by Mockapapella** [comment]
output.txt is now `true`

**Comment by tscpp** [comment]
@Seo-Rii `getNativeWindowHandle` **does** exists in Mockapapella's browser window, but `readInt32LE` does not exist in `Buffer` in his enviroment.

**Comment by Seo-Rii** [comment]
@Mockapapella It's very strange. Can you open the node at cmd and enter the code below?
`'readInt32LE' in Buffer.from('')`

**Comment by Mockapapella** [comment]
@Seo-Rii 

```
C:\Users\Quinten>node
Welcome to Node.js v12.19.0.
Type ".help" for more information.
> 'readInt32LE' in Buffer.from('')
true
```

**Comment by tscpp** [comment]
@Seo-Rii Found an issue with the `setVibrancy` function:

`win.__electron_acrylic_window__` doesn't exists because the custom `BrowserWindow` wasn't used.

https://github.com/Seo-Rii/electron-acrylic-window/blob/886a2cdc4d1a0ada0e95a39c658a563849b1f64d/src/vibrancy.ts#L216-L227

---

@Mockapapella sorry for the late response. Meanwhile, test this repro:

```js
const { app, BrowserWindow } = require("electron");
const path = require("path");
const { setVibrancy } = require("electron-acrylic-window");

const createWindow = () => {
   const mainWindow = new BrowserWindow({
      width: 800,
      height: 600,
      frame: false,
      webPreferences: {
         nodeIntegration: true,
         enableRemoteModule: true,
      },
   });

   mainWindow.loadFile(path.join(__dirname, "index.html"));

   // Fix for setVibrancy bug
   mainWindow.__electron_acrylic_window__ = {}

   setVibrancy(mainWindow, 'light');
};

app.whenReady().then(createWindow)
```

**Comment by Seo-Rii** [comment]
@tscpp I think you are right! I forgot to fix setVibrancy method after adding some function and it may caused problems. @Mockapapella can you test again with code above?

**Comment by Seo-Rii** [comment]
Solved on v0.5.5.
